### PR TITLE
xDS: add version gauge

### PR DESCRIPTION
**Pull request by junr03**
Surface the version_info field in the xDS subscriptions as a gauge in the xDS stats trees.

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
@danielhochman @htuch up for review

File: source/common/router/rds_subscription.cc [hunk 1]
```diff
@@ -50,6 +50,7 @@ void RdsSubscription::parseResponse(const Http::Message& response) {
   resources[0].set_name(route_config_name_);
   callbacks_->onConfigUpdate(resources);
   version_info_ = Envoy::Config::Utility::computeHashedVersion(response_body);
+  stats_.version_.set(HashUtil::xxHash64(version_info_));
```
Comment (danielhochman) [inline source/common/router/rds_subscription.cc]: i would just hash the response body again here

Comment (mattklein123) [reply source/common/router/rds_subscription.cc]: Please avoid computing the hash twice. Just return a tuple or something.

Comment (mattklein123) [reply source/common/router/rds_subscription.cc]: But +1 for using same hash.

File: docs/configuration/http_conn_man/rds.rst [hunk 2]
```diff
@@ -69,7 +69,7 @@ will only perform a full reload if the hash value changes.
 Statistics
 ----------
 
-RDS has a statistics tree rooted at *http.<stat_prefix>.rds.* with the following statistics:
+RDS has a statistics tree rooted at *http.<stat_prefix>.rds.<route_config_name>* with the following statistics:
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/rds.rst]: `*http.<stat_prefix>.rds.<route_config_name>.*`

**Comment by mattklein123** [comment]
Per https://github.com/envoyproxy/envoy/pull/1806 please put the RDS per table stats into a dedicated scope, so that any naming fixes are automatically picked up. cc @hennna

**Comment by junr03** [comment]
@danielhochman @htuch @mattklein123 ready for another pass.

File: source/common/config/http_subscription_impl.h [hunk 3]
```diff
@@ -105,6 +107,7 @@ class HttpSubscriptionImpl : public Http::RestApiFetcher,
   }
 
   std::string path_;
+  std::string version_info_;
```
Comment (danielhochman) [inline source/common/config/http_subscription_impl.h]: why add this local var?

File: source/common/router/rds_impl.cc [hunk 4]
```diff
@@ -38,6 +38,13 @@ StaticRouteConfigProviderImpl::StaticRouteConfigProviderImpl(
     Upstream::ClusterManager& cm)
     : config_(new ConfigImpl(config, runtime, cm, true)) {}
 
+std::string RdsRouteConfigProviderImpl::rdsStatsName(const std::string& stat_prefix,
+                                                     const std::string& route_config_name) {
+  std::string stats_name = fmt::format("{}rds.{}.", stat_prefix, route_config_name);
+  std::replace(stats_name.begin(), stats_name.end(), ':', '_');
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: Do you have a test case for this? cc @hennna please update this site also in your other PR if we move this into scope.

Comment (junr03) [reply source/common/router/rds_impl.cc]: I used all the stats expectations as a test that the name is correct. But I will make an explicit test.

File: source/common/upstream/sds_subscription.cc [hunk 5]
```diff
@@ -73,7 +73,9 @@ void SdsSubscription::parseResponse(const Http::Message& response) {
   }
 
   callbacks_->onConfigUpdate(resources);
-  version_info_ = Config::Utility::computeHashedVersion(response_body);
+  std::pair<std::string, uint64_t> hash =
+      Envoy::Config::Utility::computeHashedVersion(response_body);
+  version_info_ = hash.first;
```
Comment (danielhochman) [inline source/common/upstream/sds_subscription.cc]: no gauge?

File: source/common/router/rds_impl.cc (hunk 4 above)
Comment (hennna) [reply source/common/router/rds_impl.cc]: I've updated #1806. Let me know if I'm missing anything else.

File: docs/configuration/http_conn_man/rds.rst [hunk 6]
```diff
@@ -69,7 +69,9 @@ will only perform a full reload if the hash value changes.
 Statistics
 ----------
 
-RDS has a statistics tree rooted at *http.<stat_prefix>.rds.* with the following statistics:
+RDS has a statistics tree rooted at *http.<stat_prefix>.rds.<route_config_name>.*.
+Any ``:`` character in the ``route_config_name`` name gets replaces with ``_`` in the
```
Comment (htuch) [inline docs/configuration/http_conn_man/rds.rst]: Nit: s/replaces/replaced/

File: docs/configuration/cluster_manager/cds.rst [hunk 7]
```diff
@@ -59,6 +59,8 @@ CDS has a statistics tree rooted at *cluster_manager.cds.* with the following st
   :header: Name, Type, Description
   :widths: 1, 1, 2
 
+  config_reload, Counter, Total API fetches that resulted in a config reload due to a different config
   update_attempt, Counter, Total API fetches attempted
   update_success, Counter, Total API fetches completed successfully
   update_failure, Counter, Total API fetches that failed (either network or schema errors)
+  version, Gauge, Hash of the contents from the last successful API fetch
```
Comment (htuch) [inline docs/configuration/cluster_manager/cds.rst]: Don't we also want to report the actual `version_info` as provided by v2 xDS?

Comment (junr03) [reply docs/configuration/cluster_manager/cds.rst]: I don't think we can provide that via a gauge, given that version_info is a string;

File: source/common/router/rds_impl.cc (hunk 4 above)
Comment (junr03) [reply source/common/router/rds_impl.cc]: will remove this in favor of @hennna's change.

File: docs/configuration/cluster_manager/cds.rst (hunk 7 above)
Comment (danielhochman) [reply docs/configuration/cluster_manager/cds.rst]: we could attempt atoul on the first 6 chars similar to the server's version info and fallback to hashing. i'm not sure that it's worth it though.

**Comment by junr03** [comment]
@danielhochman updated

File: include/envoy/config/subscription.h [hunk 8]
```diff
@@ -75,18 +75,19 @@ template <class ResourceType> class Subscription {
  * Per subscription stats. @see stats_macros.h
  */
 // clang-format off
-#define ALL_SUBSCRIPTION_STATS(COUNTER) \
+#define ALL_SUBSCRIPTION_STATS(COUNTER, GAUGE) \
```
Comment (ccaraman) [inline include/envoy/config/subscription.h]: nit: align the spacing of `\`

File: include/envoy/upstream/upstream.h [hunk 9]
```diff
@@ -235,7 +235,8 @@ class HostSet {
   COUNTER(update_attempt)                                                                          \
   COUNTER(update_success)                                                                          \
   COUNTER(update_failure)                                                                          \
-  COUNTER(update_empty)
+  COUNTER(update_empty) \
```
Comment (ccaraman) [inline include/envoy/upstream/upstream.h]: nit: fix spacing of `\`

File: include/envoy/config/subscription.h (hunk 8 above)
Comment (junr03) [reply include/envoy/config/subscription.h]: hmm I thought `fix_format` would take care of this.

**Pull request merged**